File: string/check-distances-between-same-letters/check-distances-between-same-letters.go
package checkdistancesbetweensameletters

func checkDistances(s string, distance []int) bool {
	lookup := make(map[rune]int)
	for i, c := range s {
		if old, ok := lookup[c]; ok {
			if (i - old - 1) != distance[c-'a'] {
				return false
			}
		} else {
			lookup[c] = i
		}
	}

	return true
}

File: string/check-distances-between-same-letters/check-distances-between-same-letters.ts
const a_CODE = "a".charCodeAt(0);

function checkDistances(s: string, distance: number[]): boolean {
    const lookup: { [key: string]: number } = {}

    let i = 0;
    for (let c of s) {
        if (lookup[c] !== undefined) {
            if ((i - lookup[c] - 1) !== distance[c.charCodeAt(0) - a_CODE]) {
                return false;
            }
        } else {
            lookup[c] = i;
        }
        i++;
    }

    return true;
};
